torch/csrc/jit/python/python_custom_class.cpp

```diff
@@ -25,11 +25,47 @@ py::object ScriptClass::__call__(py::args args, py::kwargs kwargs) {
   return py::cast(instance);
 }
 
+/// Variant of StrongFunctionPtr, but for static methods of custom classes.
+/// They do not belong to compilation units (the custom class method registry
+/// serves that purpose in this case), so StrongFunctionPtr cannot be used here.
+/// While it is usually unsafe to carry a raw pointer like this, the custom
+/// class method registry that owns the pointer is never destroyed.
+struct ScriptClassFunctionPtr {
+  ScriptClassFunctionPtr(Function* function) : function_(function) {
+    TORCH_INTERNAL_ASSERT(function_);
+  }
+  Function* function_;
+};
+
 void initPythonCustomClassBindings(PyObject* module) {
   auto m = py::handle(module).cast<py::module>();
 
+  py::class_<ScriptClassFunctionPtr>(
+      m, "ScriptClassFunction", py::dynamic_attr())
+      .def("__call__", [](py::args args, py::kwargs kwargs) {
+        auto strongPtr = py::cast<ScriptClassFunctionPtr>(args[0]);
+        Function& callee = *strongPtr.function_;
+        py::object result = invokeScriptFunctionFromPython(
+            callee, tuple_slice(std::move(args), 1), std::move(kwargs));
+        return result;
+      });
+
   py::class_<ScriptClass>(m, "ScriptClass")
       .def("__call__", &ScriptClass::__call__)
+      .def(
+          "__getattr__",
+          [](ScriptClass& self, const std::string& name) {
+            // Define __getattr__ so that static functions of custom classes can
+            // be used in regular Python.
+            auto type = self.class_type_.type_->castRaw<ClassType>();
+            TORCH_INTERNAL_ASSERT(type);
+            auto* fn = type->findStaticMethod(name);
+            if (fn) {
```

Comment:
I think i asked this before, but does the non-static methods live in compilation unit?

Reply:
No, they also live in the TorchBind registry.

Reply:
No, they also live in the TorchBind registry.